lib/util/paging-iterator.js

```diff
@@ -0,0 +1,190 @@
+
+// -----------------------------------------------------------------------------
+// Typedefs
+// -----------------------------------------------------------------------------
+
+/**
+ * The iterator response object
+ * @typedef {Object} IteratorData
+ * @property {Array} [value] - The next set of values from the iterator
+ * @property {bool} done - Whether the iterator is completed
+ */
+
+/**
+ * Iterator callback
+ * @callback IteratorCallback
+ * @param {?Error} err - An error if the iterator encountered one
+ * @param {IteratorData} [data] - New data from the iterator
+ * @returns {void}
+ */
+
+// -----------------------------------------------------------------------------
+// Requirements
+// -----------------------------------------------------------------------------
+
+var querystring = require('querystring'),
+	Promise = require('bluebird'),
+	PromiseQueue = require('promise-queue'),
+	errors = require('./errors');
+
+PromiseQueue.configure(Promise);
+
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+const PAGING_MODES = Object.freeze({
+	MARKER: 'marker',
+	OFFSET: 'offset'
+});
+
+// -----------------------------------------------------------------------------
+// Public
+// -----------------------------------------------------------------------------
+
+class PagingIterator {
+
+	/**
+	 * Determine if a response is iterable
+	 * @param {Object} response - The API response
+	 * @returns {bool} Whether the response is iterable
+	 */
+	static isIterable(response) {
+		return response.request.method === 'GET' && Array.isArray(response.body.entries);
+	}
+
+	/**
+	 * Asynchronous iterator for paged collections
+	 * @constructor
+	 * @param {Object} response - The original API response
+	 * @param {BoxClient} client - An API client to make further requests
+	 * @returns {void}
+	 * @throws {Error} Will throw when collection cannot be paged
+	 */
+	constructor(response, client) {
+
+		var data = response.body;
+
+		if (!Array.isArray(data.entries)) {
+			throw new Error('Cannot create paging iterator for non-paged response!');
+		}
+
+		if (Number.isSafeInteger(data.offset)) {
+			this.nextField = PAGING_MODES.OFFSET;
+			this.nextValue = data.offset;
+		} else if (typeof data.next_marker !== 'undefined') {
+			this.nextField = PAGING_MODES.MARKER;
+			this.nextValue = data.next_marker;
+		} else {
+			throw new Error('Unable to determine paging strategy for response!');
+		}
+
+		this.limit = data.limit || data.entries.length;
+		this.done = false;
+
+		var queryIndex = response.request.href.indexOf('?');
+		var href = queryIndex > 0 ? response.request.href.substring(0, queryIndex) : response.request.href;
+		this.options = {
+			headers: response.request.headers,
+			qs: querystring.parse(response.request.uri.query)
+		};
+
+		// qs.parse() makes everything a string, ensure the limit is numeric
```

Comment:
There could be, but it doesn't matter because we'll just reserialize them (querystring doesn't differentiate between numeric and string types).  The reason we need to ensure limit is numeric is so when we do math with it we get number math and not string concatenation.